src/h5web/visualizations/line/DataLine.tsx

```diff
@@ -0,0 +1,46 @@
+import React, { useMemo, useCallback } from 'react';
+import { useThree } from 'react-three-fiber';
+import { Vector3 } from 'three';
+import { scaleLinear } from 'd3-scale';
+import shallow from 'zustand/shallow';
+import { useLineVisConfig } from './config';
+import { Domain } from '../shared/models';
+
+interface Props {
+  data: number[];
+}
+
+function DataLine(props: Props): JSX.Element {
+  const { data } = props;
+  const {
+    size: { width, height },
+  } = useThree();
+
+  const [abscissaDomain, ordinateDomain] = useLineVisConfig(
+    state => [state.abscissaDomain, state.ordinateDomain],
+    shallow
+  );
+
+  const xScale = scaleLinear()
+    .domain(abscissaDomain as Domain)
+    .range([-width / 2, width / 2]);
+  const yScale = scaleLinear()
+    .domain(ordinateDomain as Domain)
+    .range([-height / 2, height / 2]);
```

Comment:
The guard clause in `LineVis` ensures that `abscissaDomain` and `ordinateDomain` are always defined so `as Domain` is only to make TS happy.

Reply:
This feels a bit yuk, because it subtly couples this component to `{ yDomain && ... }` in `LineVis`.

It would be good if the cavans could be rendered without an ordinate domain and if this component could handle an `undefined` data/ordinate domain like the mesh in the heatmap. Do get there you can:

- make `axisDomain` optional in `VisCanvas` and render `AxisGrid` conditionally in the JSX;
- move the `points` computation below to a custom hook and return `undefined` if the ordinate domain is `undefined`, then render `<bufferGeometry />`  conditionally in the JSX. To go further and replicate what we do in the heatmap's `Mesh` (for performance), you could even memoise the whole material.